### ISSUE TITLE
Clarify behavior of Quantity arithmetic on shifted units

### ISSUE THREAD

**Issue by desruisseaux** [unitsofmeasurement/unit-api#95]
Clarify the behavior of `Q₁ + Q₂` or `Q * n` (with _n_ a unitless number) when the unit of measurement of `Q`, `Q₁` or `Q₂` is "shifted". Example of shifted units are:

* Temperature in Celsius, with °C defined as K - 273.15 K.
* Density sigma-t, with σT defined as ρ - 1000 kg/m³ (used in oceanography).
* Other shifted units in other scientific domains?

Taking temperature in °C as an example, addition can happen in at least two different contexts:

* T₁ + T₂ : there is some legitimate uses for such additions, but they are rare.
* T + ΔT : a more common case.

In the T₁ + T₂ case, both quantities must be converted to K before addition. So 20°C + 10°C = 303.15°C because T₂ = 283.15 K. In the T + ΔT case, 10°C is an increment (ΔT). So in this later case 20°C + 10°C = 30°C.

In current API, there is no way to know if a number is an absolute quantity or an increment. So there is no way to know if we are in the T₁ + T₂ case or in the T + ΔT case. If we want to differentiate those quantities, we may need to:

* Add a new method in `UnitConverter`: `deltaConvert(double)`, which convert the given value without applying the offset. This is similar to `deltaTransform` in `java.awt.geom.AffineTransform`.
* Introduce a new interface: `Increment<Q extends Quantity<Q>>`.
* Defines arithmetic rules as below: in any arithmetic operation like `M₁ + M₂` or `M * n`, implementation SHALL do the following steps at least conceptually (implementations are allowed to avoid conversions in some cases provided that the numerical result is the same):
  * **All** _M_ values SHALL be converted to system units before addition or multiplication.
    * Conversions of `Quantity` instances shall be done with `UnitConverter.convert(…)`.
    * Conversions of `Increment` instances shall be done with `UnitConverter.deltaConvert(…)`.
  * Result type is defined by the following table:
    * `Quantity` + `Quantity` = `Quantity`
    * `Quantity` + `Increment` = `Quantity`
    * `Increment` + `Quantity` = `Quantity`
    * `Increment` + `Increment` = `Increment`
    * `Quantity` * _n_ = `Quantity`
    * `Increment` * _n_ = `Increment`
  * Result value is converted back to the unit of measurement of the first operand (M₁).
    * Using `UnitConverter.convert(…)` if the result type is `Quantity`.
    * Using `UnitConverter.deltaConvert(…)` if the result type is `Increment`.

Requires #39
Requires #98
Requires #99

**Comment by desruisseaux** [comment]
Note: see https://rechneronline.de/temperatur/temperature.php for example of operations on temperatures. Even if the above proposal is not apply, at least we should add tests making sure that 20°C + 10°C = 303.15°C, not 30°C, unless we clearly state in the javadoc a policy about when quantities are considered increments.

**Comment by keilw** [comment]
This sounds rather intrusive for adding it to the API and introduce something like `javax.measure.Increment` or add a new method to the UnitConverter interface in that same package. Especially if it does not apply to at least 90% of all quantities (in `javax.measure.quantity` plus e.g. `si-quantity` or `systems-quantity`)
The mentioned other example `Density` is part of `si-quantity` but it does not seem used by any of the unit systems yet. 
We decided to keep `Range` or `Measurement` in implementation modules like [Indriya](https://github.com/unitsofmeasurement/indriya/tree/master/src/main/java/tech/units/indriya). Or [ComparableQuantity](https://github.com/unitsofmeasurement/indriya/blob/master/src/main/java/tech/units/indriya/ComparableQuantity.java) which in theory (Java SE8 and above now) could even be merged into Quantity, but I'm not sure about that either (will ask this in a separate ticket)
So while using `Comparable`, `Serializable` or `Prefix` apply to pretty much every quantity (maybe the only ones where `Prefix` may not make much sense are `Temperature` and a few others)
So a gut feeling tells me, the small amount of quantities (out of at least **60** in the 3 quantity modules/packages, how many are confirmed to need `Increment`?) does not justify an extra API element, but it seems reasonable to define it similar as `Range` or `ComparableQuantity`.

**Comment by desruisseaux** [comment]
I never proposed to have a method returning Quantity in `UnitConverter`. I proposed a `deltaConvert(double)` method, returning a double like the existing `convert(double)`.

The other example is not `Density`, but _sigma-t_. Density has nothing special compared to other units; it _sigma-t_ which is a shifted unit.

Yes, addition of an `Increment` interface is intrusive. But conceptually it applies to all units; not only temperature and sigma-t. It is not apparent for other units because in their case, `Quantity` and `Increment` have the same behavior.

The situation is not comparable to `Range`. Range can be defined in a totally independent way; supporting ranges in an external API do not require any change in JSR-363 API. But the situation is not the same for `Increment`. If arithmetic operations are defined in `Quantity`, then they need to define precisely their behavior with shifted units - current API is not satisfying. We can either talk about increment in the javadoc, or introduce an explicit `Increment` type.

**Comment by keilw** [comment]
The wording sounded a bit like the interface would be expected as a return value. Nevertheless if there were two methods there it had a rather significant impact on the Spec and other places to clarify how to use them and when. It is a bit like the new `fireAsync()` methods introduced with CDI 2 that are also applied and used in parallel to those already existing. The overhead in the API would not be that big, but Spec, RI and TCK would have to be changed significantly. Not entirely sure, if it had to be on the top level, but if you envision a `Quantity.add(Increment)` method then it must be, see `Prefix` and its relation to `Unit`.

**Comment by keilw** [comment]
Wonder, if someone from the SmartHome team like @htreu or @kaikreuzer might help with that at least as Contributors?

**Comment by desruisseaux** [comment]
I'm find with `convertDelta` or `convertIncrement` method names.

Introducing an explicit `Increment` interface is one possibility. Another possibility is to not introduce new interface an introduce the following methods in `Quantity` instead:

* `addIncrement(Quantity)`
* `subtractIncrement(Quantity)`
* `multiplyIncrement(Quantity)`
* `divideIncrement(Quantity)`

**Comment by keilw** [comment]
While it adds extra methods that may actually be less intrusive and complex. https://github.com/unitsofmeasurement/unit-api/issues/96 asked about possibly adding a few methods now in `ComparableQuantity` into the API. If the result of these 4 candidates is always a `Quantity`, would method overloading with `add(Quantity, FunctionType)` also work? `FunctionType` would simply be an `enum` either standalone or a `static` inner type in `Quantity`.

**Comment by keilw** [comment]
A) Add a new Increment type to the API

**Comment by keilw** [comment]
B) Reuse `Quantity` and add new methods like `addIncrement(Quantity)` or `add(Quantity, OperationType.INCREMENT)`.

**Comment by keilw** [comment]
I hope others will also cast their vote. I did here also because the whole **formatting** topic for quantities would get even more complicated if we had to differentiate between `Quantity` and `Increment` as well.

**Comment by keilw** [comment]
I know it's a little messy now, but I could update your comment and created an option C there.
Which I find even less intrusive. Let's try to call it `MeasurementLevel`, regardless of being an inner `enum` or standalone type, to better blend in with existing types like `MeasurementException`.

**Comment by filipvanlaenen** [comment]
Actually, I'd rather opt for D, don't make a change.

20°C + 10°C = 303.15°C sounds very counter-intuitive. I'm pretty sure this is going to produce a lot of bug reports.

Furthermore, I remember from thermodynamics back at university (a quarter of a century ago…) that temperature calculations were done in Kelvin, and nothing else. Not sure for physics, but I seem to remember that there too, you did the conversion into Kelvin when dealing with absolute temperatures.

I wonder whether we really have a problem for density sigma-t too, but the oceanographer would have to answer that.

Complicating factor: for integer quantity types, 20°C + 10°C will be 303°C, right?

**Comment by keilw** [comment]
I morphed your comment into option D. I take @filipvanlaenen D was your choice then?;-)

@unitsofmeasurement/experts and @unitsofmeasurement/contributors please consider voting on the three options, we have two thumbs up on the initial text, but that is not clear whether you prefer Option A, B or C to accomplish it.

**Comment by keilw** [comment]
@desruisseaux Would it be possible to vote on a choice (A-D) as well, since you presented 3 of them?
@otaviojava, @htreu would it be possible to place your thumbs-up on one of the solution paths as well? I take it you don't prefer option D (don't do it at all) but we have to chose which of the options to go for if we want to do anything at all on this.

Thanks,
Werner

**Comment by desruisseaux** [comment]
Yes, calculation in thermodynamic must be done in Kelvin. This is why 20°C + 10°C = 303.15°C if neither 20°C or 10°C are interval. Maybe 20°C + 10°C = 576.3 K would be less confusing.

Option D (no API change) would require that we choose a behavior for `Quantity.add(Quantity)` (e.g. is the argument considered an absolute measurement or an increment?) and we document it in the Javadoc. Users would need to think about whether that behavior fits his need.

I would go for a A amended by C. I.e. do everything described in A, except that the `Increment` type is replaced by a `LevelOfMeasurement getLevel()` method (maybe different name) in `Quantity`. Advantages:

* Only one method to add instead than 4.
* If in the future we discover that there is another level of measurement than we need to take in account, it would be only one enumeration value to add in `LevelOfMeasurement` instead than 8 new methods in `Quantity`.
* Avoid the need to know the level of measurement at compile-time (unless `Increment` is defined as a subtype of `Quantity`, which would be a new alternative E). Since the quantity or increment may be the result of another calculation, we may not know at compile time if the argument given to a `Quantity.add(…)` method is an `Increment` or a `Quantity`.

**Comment by desruisseaux** [comment]
Isn't too early for vote? I'm not sure that we have well explored the proposed plan and alternatives. Each comment is raising new alternatives.

**Comment by keilw** [comment]
Well we can't have too many options. I was initially worried, it could better be dealt with in a custom or implementation-specific way. And @filipvanlaenen (or what he proposed as D) seems to share that concern.
We should find out a trend (and of course like the "Compound" question those who are in favor of a certain approach also need to provide support or at least enough input on it)

**Comment by kaikreuzer** [comment]
In general, I would vote for D, i.e. do not handle increments at all.
But what I then expect is a consistent behavior, but I think for this, something needs to be changed in the implementation. What I found in https://github.com/eclipse/smarthome/pull/5697 is that `0 °C + 0 °C = 0 °C` while `32°F + 0 °C = 64 °F (=17.8 °C)`, which is simply unexpected from whatever perspective you look at it - thus I'd consider it a bug as you get different results depending on the unit that you used for the input.

> 20°C + 10°C = 303.15°C sounds very counter-intuitive. I'm pretty sure this is going to produce a lot of bug reports.

@filipvanlaenen If we do not introduce any increment handling, this would be the correct result, wouldn't it? So if you do not like that, why do you suggest to go for D?

**Comment by kaikreuzer** [comment]
> since I believe that it is the default behavior).

False assumption.
Check [this unit test](https://github.com/eclipse/smarthome/pull/5697/files#diff-4364eada58fc7651e6b4b59e35e4022eR63), which currently passes. It shows that `1 °C + 2 °C = 3 °C` (and not `276 K` as you claim). That's what I call the bug in the current implementation. It converts the second value always to the unit of the first one (instead of converting it to K) and then does the calculation. This results in the behavior that you get different results depending on the unit of your first value, which is simply wrong.

**Comment by kaikreuzer** [comment]
Just for the full picture: [This unit test](https://github.com/eclipse/smarthome/pull/5697/files#diff-4364eada58fc7651e6b4b59e35e4022eR68) actually fails, while I think that we all agree that `1 °C + 1 K` should result in `2 °C` in ANY case (treating the values as absolute Kelvin values being added as well as treating the second value as an increment).

```
Failed tests: 
  NumberExtensionsTest.operatorPlus_Quantity_Quantity_withDifferentUnit:68 
Expected: is <2 ℃>
     but: was <-271.15 ℃>
```

**Comment by dautelle** [comment]
Ahah ! It is a bug !!
In JScience we used to represent quantity internal values always in SI units (conversion was done at creation), that way no conversion is ever performed when adding/multiplying quantities (quantities values are added or multiplied together).

**Comment by kaikreuzer** [comment]
> Ahah ! It is a bug !!

Just my words 😉

**Comment by desruisseaux** [comment]
I don't think that 1°C + 1 K = 2°C in any case. It depends if 1°C is an absolute temperature or an increment. We could decide that in A + B, the first operand (A) is an absolute temperature and the second operand (B) in increment, but this is an arbitrary choice.

Option D seems dangerous to me because it relies on convention (e.g. first operand is assumed absolute and second operand is assumed increment). If A or B is the result on another calculation, it can be anything. We expose the users to somewhat random behavior depending on the order in which he applies his operations. For example if A, B and C are absolute temperatures, then the behavior of the `add` method in `A.add(B)` should be different than the behavior of the same `add` method in `A.add(B.subtract(C))`. If we want to provide correct answer in those two cases, then we need to know if a quantity is an increment or not (either with subtypes or with enumeration values). For example `B.subtract(C)` would return a `Quantity` with `LevelOfMeasurement` set to interval, so that `A.add(…)` can use this information for switching to the proper behavior.

**Comment by dautelle** [comment]
Actually, the meaning of the quantities is up to the user. 
You can do  B + A (with B being an increment and A the increment above the absolute zero), it will lead to the same result. If a user wants to hold a variable which is a temperature above the melting point of lead, why not! If he adds two such variables together it will not make senses.

**Comment by keilw** [comment]
@dautelle Can you point to where JScience does or did this?
While there are many often patched versions of JScience 4 all over GitHub, this https://github.com/javolution/jscience must be the `SNAPSHOT` version of the recent codebase. Except an abstract base class `Amount` I did not find anything similar to the JSR 363 `Quantity`.
JSR 363 does not store everything in SI units when a `Quantity` is created. If you do A+B then B is converted into A as long as the quantity type is compatible.

**Comment by dautelle** [comment]
Yes it is the Amount classs (http://jscience.org/api/org/jscience/physics/amount/Amount.html)

**Comment by desruisseaux** [comment]
@dautelle yes, the meaning of the quantities is up to the user. Which is why we need either a way to tell to `Quantity.add(…)` how to interpret the quantity (options A, B and C), or adopt a convention based on argument order (my interpretation of option D).

**Comment by kaikreuzer** [comment]
> Option D seems dangerous to me because it relies on convention (e.g. first operand is assumed absolute and second operand is assumed increment).

I understood option D as not supporting increments at all, but treating all values as absolutes.
I thought that A+B were about introducing increment handling.

Whatever: I still vote for option X, which is to treat all as an absolute, which would require to do all temperature calculations only after converting all values to K.

**Comment by dautelle** [comment]
Agree with option X, but considering all quantities are increment (which can be an increment above the absolute zero == absolute value).

**Comment by desruisseaux** [comment]
Treating all values as absolute: 1°C + 2°C = 276.15°C.
Treating all values as increment: 1°C + 2°C = -270.15°C

**Comment by desruisseaux** [comment]
Ah sorry, my last comment is wrong. 1°C + 2°C = 3°C since the result is also an increment.

**Comment by keilw** [comment]
In this GitHub mirror: https://github.com/tarelli/jscience/blob/master/src/org/jscience/physics/amount/Amount.java I cannot see any explicit conversion from the given Unit into a SI unit. It is possible, because JScience does not support more than what JSR 275 did back then, there are mostly SI units available to use, but both `CELSIUS` and `KELVIN` were defined in https://github.com/tarelli/jscience/blob/master/src/javax/measure/unit/SI.java (beside several non-SI units like `BIT` or `BYTE` ;-)

**Comment by dautelle** [comment]
https://github.com/javolution/jscience/blob/master/physics/src/main/java/org/jscience/physics/amount/Amount.java

**Comment by dautelle** [comment]
Note: Implementation is not on GitHub yet (migration pending)

**Comment by keilw** [comment]
I would say X is up to a consuming system like SmartHome or others, I would not force even the RI to do this under every circumstance. 
What I could imagine, that's why I mentioned Indriya https://github.com/unitsofmeasurement/indriya/blob/master/src/main/java/tech/units/indriya/function/QuantityFunctions.java, is something like an `addConverted(Unit targetUnit, Quantity... quantities)` method that would add up two or more quantities after explicit conversion. In theory that could also be something like `Quantity.add(Quantity, Unit)` in the API.

**Comment by keilw** [comment]
Implementation of JSR 363? 
I still haven't heard from Harold about our demand for a list of compatible implementations at JCP.org (similar to what e.g. 107 got) but if the JCP does not have the resources for that any more, we could still use the UoM GitHub site.

**Comment by filipvanlaenen** [comment]
A few more comments:

- I think there's a problem with option C. I think that the level of measurement is a property on the unit, not the quantity.
- Without having checked it, what is the meaning of a millidegree Celsius? As far as I understand it, officially 1 m°C = 1 mK, not 273,151 mK or 273.15 mK. We should check what our implementation says.
- This also means that 1 °C + 1 m°C = 1.001 °C (i.e. 273.151 K), and not 274.151 °C.

I'm still there that people who handle temperatures should know there stuff. Adding temperatures doesn't make sense, just like adding dates doesn't make sense, unless you're trying to calculate an average. Therefore, I think 1 °C + 2 °C should be 3 °C.

Should we also consider the behaviour of logarithmic units if we're going to change the API?

That being said, if the choice is between A, B and C, I'm leaning towards A.

**Comment by keilw** [comment]
Of you use the sum(Unit) Lambda in uom-se as mentioned above (it was there already, Indriya has not added it) then you can take a `List` or array of `1 °C, 1 °F, 1 K` and sum them all in `K` (or another matching unit) and do the same for any given order of quantities. The result must be the same.

**Comment by desruisseaux** [comment]
Adding temperature makes sense in thermodynamics: this is equivalent to adding kinetic energy (scaled by _Boltzmann constant_ factor). Furthermore the problem is not only about addition, but also multiplication. What is the result of 2 * 10°C? [This web site](https://rechneronline.de/temperatur/temperature.php) said _"The common belief, that twice 10°C (or °F) equals 20°C (or °F), is wrong"_. Actually it depends if 10°C is an absolute temperature or an increment.

If we adopt a convention, then there is the consequences:

* If we said _"all operands are absolute temperatures"_, then 10°C + 20°C or 2*10°C will give surprising results for non-scientific users.
* If we said _"all operands are increments"_, then we have an inconsistency between the meaning of operands in arithmetic operations and the meaning of arguments in `Unit.getConverterTo(Unit)`, because the later handles temperatures as absolute. If addition operands are increments, than we should have 2°C + 3°C = 5°C = 5 K, but the Unit API will give us 2°C + 3°C = 5°C = 278.15 K, which is wrong in this context.
* If we said _"the first operand is absolute temperature and the second operand is increment"_, then 2°C + 3°C = 5°C = 278.15 K is consistent. But we get a non-commutative addition where A + B is not always equal to B + A (the non-commutativity aspect appears when A and B are not in the same units of measurement).

I don't think there is a safe, consistent, non-ambiguous way to perform the correct calculation without knowing if the value is an absolute measurement or an increment. It can be a `Quantity` property, a `Unit` property or a `Quantity` subtype. Jean-Marie's proposal to make it a `Unit` property instead of `Quantity` seems good to me.

**Comment by desruisseaux** [comment]
@keilw the fact that `testSumTemperatureK2C` verifies that 1°C + 1 K = 2°C does not mean that this is the right thing to do. As explained above, it depends on what 1°C and 1 K are. If we don't specify that, you will always find users saying that 1°C + 1 K = 2°C is wrong.

**Comment by keilw** [comment]
@desruisseaux @dautelle Can you explain what

> Jean-Marie's proposal to make it a Unit property instead of Quantity seems good to me.

Should look like?

**Comment by dautelle** [comment]
As always there is no definite "right" answer to the problem.  But there are somes properties which have to be respected: 
  1.  Temp.a + Temp.b = Temp.b + Temp.a (commutativity)
  2.  Temp.a.inAnyUnit + Temp.b.inAnyUnit = Temp.a.inAnyOtherUnit + Temp.b.inAnyOtherUnit 

Since we add quantities, the unit should be irrelevant. The real question is:
    What quantity is 1°C ?
Is it equal to 1 °K or 274,15 °K ?
If you consider that it is 1°K then a statements such as "the temperature is 20°C today" is wrong (unless you are located on Pluto surface ;) 
If you consider that it is 274,15 °K then a statements such as "the temperature tomorrow is 5°C greater than today" is wrong (we are not on Mercury surface either ;)
What ever the choice, there will be confusion. 
What is the least objectionable solution? I believe it is the second one which corresponds better to the physical quantities (kinetic movement). The statement "the temperature tomorrow is 5°C greater than today" will be incorrect, the correct statement will be "the temperature tomorrow is 5°K greater than today". In other words when dealing with delta/differences the correct unit should always be °K

**Comment by keilw** [comment]
JSR 363 is Final for almost 2 years and was available as Draft about 1 1/2 years longer, and I don't remember a single issue ticket saying "1°C + 1 K = 2°C is wrong, please fix it".  
Actually until the SmartHome question we never even received an issue saying "1 °C + 1 K = -271.15 °C is problematic for my use case, please fix it". 

So not sure, if we shall add complexity to the API and cause all sorts of headaches when dealing with "Compound" formatting (if we even do in 2.0?) for rare corner cases. Allowing to control which target unit is used for a particular operation can solve the "1 °C + 1 K vs. 1 K + 1 °C" dilemma.

**Comment by desruisseaux** [comment]
Possible approach: add the following methods in `Unit` interface:

* `LevelOfMeasurement getLevel()` where `LevelOfMeasurement` is an enumeration with `NOMINAL`, `ORDINAL`, `INTERVAL` and `RATIO` values (only `NOMINAL` and `ORDINAL` could be omitted for now). See https://en.wikipedia.org/wiki/Level_of_measurement
* `Unit<Q> level(LevelOfMeasurement)` for returning the same unit of measurement than `this`, but using a different level of measurement. So if `CELSIUS` represents a unit of temperature, then `CELSIUS.level(LevelOfMeasurement.INTERVAL)` returns a unit of temperature _increment_.

Then, update `Unit.getConverterTo(Unit)` contract as below:

* The two units shall have the same `LevelOfMeasurement`.
* If level of measurement is `RATIO`, then return a `UnitConverter` as today (e.g. 10°C is converted to 283.15 K or 50°F).
* If level of measurement is `INTERVAL`, then return a `UnitConverter` that ignore the offset (e.g. 10°C is converted to 10 K or 18°F).

For implementation of `Quantity` arithmetic operations, if the right converter is used, correct calculation should follow automatically. Example (omitting parameterized types for simplicity):

```
Quantity add(Quantity that) {
    Unit u1 = this.getUnit();
    Unit u2 = that.getUnit();
    Unit uc = u1.level(u2.getLevel());    // We will convert in unit of u1, but taking in account the nature of u2 (quantity or increment).
    UnitConverter c = u2.getConverterTo(uc);
    newValue = this.getValue() + c.convert(that.getValue());

    // The result is in unit of u1, but is it an absolute value or an increment?
    // Note: following code could be factorized in a convenience method.
    boolean isIncrement1 = u1.getLevel() == UnitOfMeasurement.INTERVAL;
    boolean isIncrement2 = u2.getLevel() == UnitOfMeasurement.INTERVAL;
    boolean isResultAnIncrement = u1 & u2;
    Unit uf = u1.level(isResultAnIncrement ? UnitOfMeasurement.INTERVAL
                                           : UnitOfMeasurement.RATIO);

    return new Quantity(newValue, uf);
}
```

Yes, I agree it looks complicated. But actually I think it only adds one line of code in arithmetic method implementations if the code computing `uf` is factorized in a convenience method. If we don't do that, then the complexity will be in the Javadoc instead than in the code since we will have to specify a behavior based on conventions, which I think may be worst.

**Comment by desruisseaux** [comment]
@keilw this whole thread started from a bug report in adding temperature values. Current behavior is mathematically inconsistent, or at least ambiguous. "1°C + 1 K" must be equal to "1 K + 1°C" - unless we throw away fundamental mathematical properties like commutativity. And what should be "1°C + 1°F"? 

We can not escape this complexity - either we augment the API, or either we put the complexity in the Javadoc. I think the later is more dangerous because the users would have to remember a set of arbitrary conventions when using the Unit API.

I think that this issue should have priority over compound units because it touches mathematical properties, while compound units (as I see it - I may be wrong) seems mostly a formatting issue.

**Comment by dautelle** [comment]
On that point I think I agree with Werner (sorry Martin ;). I prefer to see in the documentation a note on Celsius/Fahrenheit units stating that these units should never be used for temperature increments but only for absolute temperatures rather than adding new methods to the API.

**Comment by keilw** [comment]
The issue reported by SmartHome can be solved if you use the `sum(Unit)` function as opposed to simply calling `sum()`. The default behavior of chained `Quantity.add()` methods would be to convert every subsequent argument into the first unit. We could expose the control the RI Lambda gives by allowing a "target unit" also to be passed in `Quantity.add()` as a second argument, that is all I could see if we wanted to offer such control outside the Lambda call, please let's create a more distinct issue (this one is a bit too broad) here, otherwise most work would be in a guidebook or similar, potentially a bit in the spec (when it comes t general existence of interval vs. absolute) but no further API elements.

**Comment by desruisseaux** [comment]
@dautelle:
>  I prefer to see in the documentation a note on Celsius/Fahrenheit units stating that these units should never be used for temperature increments but only for absolute temperatures

Then we would accept that 1°C + 2°C = 276.15°C and 2*10°C = 293.15°C? If so, it should be clearly stated in the Javadoc and verified by the conformance test suite. Note that it applies not only to temperature, but *all* units with an offset (e.g. sigma-t and probably some others).

@keilw:
> The default behavior of chained Quantity.add() methods would be to convert every subsequent argument into the first unit.

So we would break addition commutativity rule? 1°C + 1 K would not be equal to 1 K + 1°C.

**Comment by keilw** [comment]
@desruisseaux Yes, the default behavior must be backward compatible with the 1.0 version.
Due the special nature of the conversion right now they are not the same.

**Comment by desruisseaux** [comment]
@keilw: why talking about target units? It is unrelated to this problem.

Unit API 1.0 version said nothing about how arithmetic operations should be done, and RI is not a reference if its behavior is incorrect.

**Comment by keilw** [comment]
It solves the issue @kaikreuzer has that lead to creating this discussion. 
> Whatever: I still vote for option X, which is to treat all as an absolute, which would require to do all > temperature calculations only after converting all values to `K`.

That calculation to `K` is exactly what the `sum(Unit targetUnit)` function in the new RI (and uom-se, there are naturally no Lambdas in the previous RI) accomplishes. I don't see why we must also add it to the `Quantity` interface. And all who made a clear vote here (including 2 Spec Leads) voted on either **Option D** (no changes to the API) or "X" (ability to convert to a distinct unit like `K` before an operation)

Nothing prevents a implementations like JScience or SIS from adding such feature, but the RI won't support it as there was no demand for it so far. And it only seems relevant for exactly one quantity that's  already defined. Others do not exist not even in extension modules at least none we know of.

**Comment by desruisseaux** [comment]
Werner, I don't care about those votes. Even if there is 10000 votes for a mathematically broken feature, it is still broken. One million votes will never change that.

> That calculation to K is exactly what the sum(Unit targetUnit) function in the new RI

Okay. So the specification could be: _"All arithmetic operations must be done as if all units were converted to system unit before the operation. Implementation are allowed to take shortcut (e.g. avoid some conversions) provided that the numerical result is the same, ignoring rounding error."_ Is that specification okay?

If we accept the above, do we agree that it implies that 1°C + 2°C = 276.15°C and 2*10°C = 293.15°C?

**Comment by keilw** [comment]
Why the heck would 1°C + 2°C = 276.15°C ? I think @filipvanlaenen mentioned we would get all sorts of issues and bug tickets if that was the case.

**Comment by desruisseaux** [comment]
The link that you gave tell it: _"You need to distinguish between a temperature measurement and a temperature difference."_ — this is the topic of this thread since the beginning. If the group votes for not distinguish them, then we have be specific about what the arithmetic operations must do.

Jean-Marie suggested to decide that all operands are measurements (not differences). You suggested that all temperatures are converted to Kelvin before addition — which is a consistent approach. A consequence of those two suggestions is that 1°C + 2°C = 274.15 K + 275.15 K. Yes it will cause confusion for some users. But this is math. The fact that the Earth is round too confuses some peoples. The Earth will not become flat because some peoples "vote" for it.

If you don't like the above and want 1°C + 2°C = 3°C, then we have to admit that at least one of those two quantities is not a measurement but a difference (or increment). This is fine. But then we need to specify the rules in the Javadoc, otherwise we will get random behavior depending on implementations. There is some possible rules:

1. The first operand is a measurement and the second operand is an increment. **Problem:** break addition commutativity (A + B ≠ B + A).
2. The first operand is an increment and the second operant is a measurement. **Problem:** same than above.
3. The two operands are increments. **Problem:** calculation done in different units produce inconsistent results: if A and B are in °C, (A + B) converted to K is inconsistent with (A converted to K) + (B converted to K).
4. The two operands are measurements (Jean-Marie's proposal): mathematically consistent, but un-intuitive for users not familiar with thermodynamic.
5. Convert second operand in the unit of first operand (Werner's proposal): this is equivalent to alternative 2 above.

So what is your choice?

**Comment by keilw** [comment]
With `Kelvin` as a result it makes more sense.

5. is not my proposal but the default behavior of implementations. Since Java SE 8 implementations the alternative exists which allow to define the desired unit regardless of the order (thanks @otaviojava btw for introducing it ;-)

**Comment by keilw** [comment]
This C# project won't work in Visual Studio unlike others, but it has a Wiki explaining the special case for Temperature (it is not used anywhere else) https://github.com/mangh/unitsofmeasurement/wiki/Quantities-vs.-levels,-units-vs.-scales
Calling it `level` / `scale`.
As `Level` is a quantity type in the ISQ for `BEL` or `DECIBEL`, I'd try to avoid it. 

Again I am not sure if we need that, and for a single quantity being applied by other libraries that even bother to distinguish.

https://www.graphpad.com/support/faq/what-is-the-difference-between-ordinal-interval-and-ratio-variables-why-should-i-care/ mentions one other case which is Acidity. @magnonasc added it to the systems-quantity module, but I don't think it is used yet.
However, if we look at UCUM: http://unitsofmeasure.org/ucum.html
which is also used by an extension module, how many units there would fall into a certain category?

These other sources
- https://www.spss-tutorials.com/measurement-levels/
- http://www.statisticshowto.com/ratio-scale/

Are interesting, because they distinguish between those MeasurementLevels in the context of statistics and data analytics.
Given the Physikal by @ejberry is also used in data analytics and smart energy use cases, I wonder, what he thinks about the matter? If a simple enum like MeasurementLevel helped to make decisions also for statistical analysis, then it may be more beneficial than it does harm or add complexity. 
If we do, I'd even declare all of them but leave it to implementations how many are used. `Nominal` and `Ordinal` may not be so common with numeric values, but use cases like the T-Shirt sizes in https://github.com/unitsofmeasurement/uom-impl-enum or Agile story points could be examples of using them as well, so I would not exclude them.

Thoughts?

**Comment by keilw** [comment]
Although it's almost more implementation specific, something along those lines could be under design goals in the spec. 
Adding an `Interval` or `Scale` type to the API for still a very small number of actual units would be toxic. And complicate other aspects especially formatting. 
I'd be fine with either **C** (adding a `MeasurementLevel` to wherever it makes more sense, probably the `Unit`) or **D** (no change to the API but recommendations in Spec, UOM Guidebook or JavaDoc)

**Comment by keilw** [comment]
So far after cleaning some duplicates (please try to only vote for one option each) there's
A) 1? (@filipvanlaenen mentioned leaning towards A despite having brought up D)
B) 0
C) 2 (as @desruisseaux brought up at least 3 options, could you decide for one please?)
D) 1 Although @filipvanlaenen brought it up, he mentioned a tendency towards A, so I count only @dautelle's thumb up.

Option X can be accomplished by using the right Lambda for Temperature calculations, therefore it does not count here, if possible could anyone who has not stated his preference do so please. (@otaviojava, @htreu who placed a thumb on the header, so it cannot be told to a particular option, @kaikreuzer if you prefer one of the 4 options or @ejberry if you prefer either of them for use by Kotlin and Physikal. As well as all others if you can)

Thanks,
Werner

**Comment by filipvanlaenen** [comment]
My preference vote is as follows:
1) D (a.o. because I think we have to think this more through, and address m°C and logarithmic units)
2) A
3) C (how do you parse “1 °C” to the correct level of measurement, because 1 °C as a temperature increment is allowed and equal to 1 K)
4) B

**Comment by keilw** [comment]
Thanks, for the sake of simplicity I only count your first preference and updated the summary above.

**Comment by keilw** [comment]
Well this is the wiki, and those who prefer not to change the API currently have the most votes.

The Wiki for that C# template: https://github.com/mangh/unitsofmeasurement/wiki/Quantities-vs.-levels,-units-vs.-scales only covers `Temperature` for the `scale` type.

It is implementation specific and if someone took the time to write JUnit tests for every single quantity, it'll show, that for every quantity type except `Temperature` `Quantity A.add(B) `is the same as `B.add(A)`.

For the special case of `Temperature` additions explicit conversion to the system unit `K` is accomplished through the Lambda `sum()`. The JUnit `testSumTemperatureEquality` in https://github.com/unitsofmeasurement/indriya/blob/master/src/test/java/tech/units/indriya/function/QuantityFunctionsTemperatureTest.java (I called `FunctionsTemperatureTest` now to point out its purpose) demonstrates, `1 °C + 1 K = 1 K + 1 °C`.  The QuantityFunctions class does not use any RI classes, it could even enter the API as the next version relies on Java SE 8 and above.

**Comment by desruisseaux** [comment]
> It is implementation specific and if someone took the time to write JUnit tests for every single quantity, it'll show, that for every quantity type except Temperature Quantity `A.add(B)` is the same as `B.add(A)`.

Werner, it is the case for **every units having an offset**, not just temperature. It will be the case for density sigma-t, for pH and all other domains that we don't know.

The fact that the problem show up with temperature **prove that there is a mathematical problem**. One million on votes will not change that. If you decide to note change the API, you still have to define precisely the operation behavior.

Please stop talking about lambda functions, it is irrelevant to this discussion! We are talking about mathematical definitions. Only after we have decided what should be the math, we can discuss how to translate that in an API or in Javadoc.

Do you think that JCP will accept a specification that pretend that the result of A + B is implementation details?

**Comment by desruisseaux** [comment]
Werner, I remind you that many of the reasons why JSR-275 has been rejected were the same than the objections I did in the expert group - the JCP and me were in agreement. Again, I told you today that it is unacceptable to have a specification that leave the result of "A + B" as an implementation details, and I will have no hesitation to recommend the JCP to vote "no" to this JSR on this basis.

**Comment by desruisseaux** [comment]
Added separated issues for the mathematic part: #98 and #99. Lets do things in order: mathematic first, then the choice between API changes or Javadoc clarification only after we agree on the math.

**Comment by dautelle** [comment]
> 1 ℃ + 1 K = (-271.14999999999992 ± 5.7E-14) ℃
> 1 K + 1 ℃ = (275.14999999999992 ± 5.7E-14) K

Are both valid calculations since we can assume that the delta is in Kelvin.
But the result is not correct since  1 ℃ + 1 K should be equal to 1 K + 1 ℃ 
The correct answer is:
> 1 ℃ + 1 K = 2 ℃
> 1 K + 1 ℃ = 275.15 K

What is not correct (and is about the  same as trying to add dates together) is:  1 ℃ + 1 ℃ 
To avoid any confusion the specification should state that adding offset units together leads to unspecified behavior (and obviously it is not recommeded). If there is a mistake in the test suite it should be corrected.

**Comment by keilw** [comment]
That's a good point. I'm not sure, if #98 and #99 are concrete action items, but I marked both as required (or blockers) to this issue in Waffle.

**Comment by keilw** [comment]
@andi-huber was the driving contributor behind https://github.com/unitsofmeasurement/indriya/wiki. One such page about Unit Equivalence could potentially also have some reference here. While it seems better to explain things in a Wiki after the JSR had some release (EDR1 should not be too long, regardless of how many additional features are ready, it is still a Draft) feel free to create a similar Wiki here. We may always relocate content e.g. to a Spec or other places later. 

Simply declaring `1°C + 1°C` as invalid in the Spec or a similar place (including JavaDocs) would be OK, but without proper meta-information like the mentioned `MeasurementLevel` or something similar I would very much refrain from throwing a `RuntimeException` in such case. Users of JScience 4, JSR 363 or the intermediary solution uom.org all of which give the same result here should not run into unpredicted behavior. Whatever issue may exist for a small amount of quantity types, all of these libraries have a massive user base and used them for a long time now. If this was JIRA the extra tickets should be sub-tasks, so if others are created, please try to connect them via "Requires" in the first description field. @dautelle already confirmed dates in the Doodle: https://doodle.com/poll/gvnu2sh7wzk4yp8t Please everyone especially those vocal here try to also find a slot. To discuss the options and who may be doing what as a result.

**Comment by desruisseaux** [comment]
> Simply declaring 1°C + 1°C as invalid in the Spec or a similar place (including JavaDocs) would be OK,

Before to do that, I want to expose the mathematical consequences of that approach. Before to expose the mathematical consequences, I want to ensure that we agree on a requirement to follow arithmetic laws, i.e. #99.

I also want to insist that all the discussion that I want to do is about **math**, not opinions.

**Comment by andi-huber** [comment]
Let me refine on @desruisseaux requirements from above:

**Scenario 1 - Calculus with Quantities of a single Unit**

Let A, B and C be Quantities of Unit 'CELSIUS', and for a moment forget about any physical interpretation, then simple arithmetic trivially fulfills these rules
* Addition commutativity: A + B = B + A
* Addition associativity: (A + B) + C = A + (B + C)
* Multiplication associativity: (A * B) * C = A * (B * C)
* If A = B (same units), than A + C = B + C and C + A = C + B
* If A = B (same units) and n is a scalar, then n*A = n*B.

Meaning 1°C + 2°C = 3°C and 10 * 3°C = 30°C ... is consistent math (again leave any physical interpretation aside).

**Scenario 2 - Calculus within Quantities of different Units**

Now let's have a closer look on what happens, when we pull Unit-Conversion into the picture. I believe, there is nothing that breaks the math ...

Let A be a Quantity of Unit 'CELSIUS' and B be a Quantity of Unit 'KELVIN', then we can state the following:
* Addition commutativity: 
```
    convertToKelvin(A) + B = B + convertToKelvin(A)
    A + convertToCelsius(B) = convertToCelsius(B) + A
```

I'm leaving out the other rules from above, since I hope its easy to see, that these are equivalently fulfilled. Unit-Conversion does not break the math.

So for mixed unit addition:

1°C + 1K ... is either 
    `1°C + convertToCelsius(1K) = 2°C`
or
    `convertToKelvin(1°C) + 1K = 275.15K`

whatever result unit desired, where again either way the following both hold true:
```
    1°C + convertToCelsius(1K) == convertToCelsius(convertToKelvin(1°C) + 1K)
    convertToKelvin(1°C + convertToCelsius(1K)) == convertToKelvin(1°C) + 1K
```


**Conclusions**
	
I guess this is the current state of the API and its perfectly consistent regarding math. And as long as you leave any physical interpretation aside (stay in the pure math picture), its also perfectly sane to state:

20°C + 5°C := 25°C ... that's the interpretation that allows for calculating a mean temperature as well as adding delta temperatures; no worries there!

**Comment by desruisseaux** [comment]
Thanks you very much @andi-huber for focusing on the math!!!!

> 1°C + 1K ... is either
> 1°C + `convertToCelsius(1K)` = 2°C
> or
> `convertToKelvin(1°C)` + 1K = 275.15K

Above equations have an error: `convertToCelsius` is defined without the 273.15 offset (so that 1°C = 1 K) while `convertToKelvin` is defined with that offset (so that 1°C = 274.15 K). In other words, `convertToCelsius(convertToKelvin(A)) ≠ A` in above equations.

Furthermore lets look closer at the 20°C + 5°C case in the context of the following arithmetic law: (A + B) + C = A + (B + C)

(20°C + 5°C) + 1K
`convertToKelvin(20°C + 5°C) + 1K` =
`convertToKelvin(25°C) + 1K` = `299.15 K`

But if we apply associativity:
20°C + (5°C + 1K)
`20°C + (convertToKelvin(5°C) + 1K)` =
`20°C + 279.15 K` =
`convertToKelvin(20°C) + 279.15 K` = `572.3 K`

So `Quantity` additions defined that way would not follow addition associativity (unless we add physical consideration in `convertToKelvin`, but I fully agree to leave physical interpretation aside and stick to pure math at this stage). If we trig unit conversions only when the two operands have different units, then rearranging terms e.g. from (A + B) + C to A + (B + C) will trig different units conversions with different side effects.

**Comment by andi-huber** [comment]
Thanks @desruisseaux !
> convertToCelsius is defined without the 273.15 offset
Sorry, for not being explicit about this: When writing up my argument, I had assumed convertToCelsius() to include the offset.

Otherwise of course the rules break down.

**Comment by desruisseaux** [comment]
> Sorry, for not being explicit about this: When writing up my argument, I had assumed `convertToCelsius()` to include the offset.

But then, 1°C + `convertToCelsius(1K)` = 1°C + -272.15°C = -271.15°C (if Celsius degrees are added directly - which as said before would be another discussion).

**Comment by andi-huber** [comment]
Yes, you are totally right, I need to correct my statement!

**Comment by filipvanlaenen** [comment]
I've been thinking about a solution that may leave things as they are today for the vast majority of units, and at the same time handle temperatures and any other units with offsets through subclassing or a an new property on `Quantity`.

The idea is that for quantities for which it doesn't matter whether they are absolute quantities or increments, calculations stay as they are today. However, if there is a difference, the calculations throw an exception, and should be performed with an explicit context of whether the quantity is an absolute one or an increment.

E.g. (suffix a denotes absolute quantities, suffix i denotes increments, no suffix a general quantity)
1 Ωa + 2 Ωa = 3 Ωa
1 Ωi + 2 Ωi = 3 Ωi
1 Ωa + 2 Ωi = 3 Ωa
1 Ωi + 3 Ωa = 3 Ωa
so therefore
1 Ω + 2 Ω = 3 Ω

However:
1 °Ca + 2 °Ca = 276.15 °Ca
1 °Ci + 2 °Ci = 3 °Ci
1 °Ca + 2 °Ci = 3 °Ca
1 °Ci + 2 °Ca = 3 °Ca
so therefore
1 °C + 2 °C = undefined
but
1 °C + 2 °Ci = 3 °C

Figuring out whether a calculation can be done should be hidden in the logic inside, but basically boils down to checking whether the converter contains a shift operator.

Notice though that a multiplication operator on a shift operator eliminates the shift operator. E.g.
1 m°C + 2 °mC = 3 m°C

**Comment by andi-huber** [comment]
Thanks @desruisseaux , I've corrected my long argument above, which also renders my original point somewhat useless. However, (and I guess one of your concerns) the following does not hold true, but also is not part of the required rules:

`convertToKelvin[20°C] + convertToKelvin[5°C] == convertToKelvin[25°C]`

**Comment by desruisseaux** [comment]
Hello @filipvanlaenen. Thanks a lot for your comment. Yes, I agree with your proposal - it is in the spirit of the initial proposal of this #95 issue. Before to discuss it, can we check that we are in agreement with the following steps please, or if any of those steps require further discussion?

1. **The specification shall define arithmetic operations with enough details for ensuring determinist results** (ignoring rounding errors or change of units) - #98
2. **Arithmetic operations shall obey to arithmetic laws** (ignoring rounding errors or change of units) - #99
3. **Current implementations do not obey to arithmetic laws** (except when the units is only a ratio, without offset). _This is the recent discussion with inputs from @andi-huber. If I'm wrong, I can be convinced by a mathematical demonstration._ Each of us can provide equations demonstrating their point of view, and other peoples try to identify errors in those equations. When we reach correct equations, we should have a clear "yes" or "no" answer to the question about whether current behavior is consistent or not.
4. **The API can be made determinist and compliant with arithmetic laws in two ways:**
   * **Javadoc only:** a solution exist (that would be the topic of another discussion) at the cost of surprising behavior for some users.
   * **With new API:** as suggested by @filipvanlaenen, and also as suggested by myself since the beginning of this thread. Details of this approach would be another discussion.

@keilw: I will accept group decision about **HOW** to make the API determinist and compliant with arithmetic laws (Javadoc or API changes). What I'm fighting for is the recognition that current behavior is **not** consistent with arithmetic laws, that arithmetic consistency is a must, that not all proposals in this thread fix the arithmetic consistency problem, and that whether a proposal is mathematically consistent or not is not a matter of opinion.

**Comment by desruisseaux** [comment]
Thanks @andi-huber - the new description looks good. As said in your discussion, 1°C + 1 K can produces 4 different solutions. Do we have an agreement in this group that a simple rule like _"convert second operand in the unit of the first operand"_ have the following problems when different "shifted units" (temperature, sigma-t, …) are mixed?

* A + B ≠ B + A
* (A + B) + C ≠ A + (B + C)

Assuming that #99 is accepted, my proposal is to discard the _"convert second operand in the unit of the first operand"_ rule on the basis of arithmetic laws violation. But we can explore other rules if the group wishes?

> However, (and I guess one of your concerns) the following does not hold true, but also is not part of the required rules:
>
> `convertToKelvin[20°C] + convertToKelvin[5°C] == convertToKelvin[25°C]`

Indeed, I would expect A + B to produce equivalent result (after unit conversions) no matter the units of A and B. But this statement is outside the arithmetic laws, so I'm less sure about that. I suggest to focus on arithmetic laws for now, and revisit above statement later if needed.

**Comment by dautelle** [comment]
A more general rule (applicable to all arithmetic operations or functions) 
    "All operations on quantities are performed on their values stated in their system unit (SI)." 

Note: This is the way JScience used to work (internal amount in SI unit, the quantities unit is only used to format /parse quantities). That way unit conversion needs to be performed only during input/output and never during internal processing.

**Comment by desruisseaux** [comment]
@dautelle: YES!!! This is exactly one of my proposals earlier in this thread! With an amendment saying "implementations are allowed to avoid conversions (as an optimization) if the numerical result is the same".

However do we agree that a consequence of this rule is that 1°C + 2°C = 276.15°C (if this operation is allowed)? Whether this operation should be allowed or not is a separated discussion. Note also that @filipvanlaenen proposal can help to solve this issue, but let keep whether to accept API changes or not in a separated discussion.

**Comment by keilw** [comment]
@dautelle Not sure which version, but the one that was based on JSR 275 did not, see the examples.
It would be rather cumbersome and require a lot of explanation if the default `Quantity.add()` method (unless e.g. it's assisted by that new property @filipvanlaenen and others hinted) forced every operation into a system unit. Without the proper context even `1 mile + 1 inch` would by default end up as `metre`, not what people expect in that case. And forcing them to convert it back to `mile` or `inch` again is cumbersome and counterintuitive. 
Whatever mathematical principles we try to obey, an intuitive API that is user-friendly is probably as important and so is a certain level of predictability, e.g. not suddenly change the behavior for every single quantity type because a handful need a different approach.

**Comment by desruisseaux** [comment]
@keilw: the proposal is not that the *result* must be in system units. The proposal is that the *calculation* is performed in system units. The result of `1 mile + 1 inch` can be in whatever units the implementation wants, but the result must be **as if** the implementation has converted everything to meters, performed the calculation, then converted back the result to mile (or inch). The implementation does not need to _really_ do all those conversions, but the result must be the same (ignoring rounding errors).

Also please focus on math, not on what JSR 275 did.

**Comment by keilw** [comment]
@desruisseaux As long as there is no massive trade-off doing multiple conversions, it could be done in some cases, but a "target unit" then must be kept trace of, even if that unit is not given as an argument to a new/updated method (which could be a possible API change, similar to what the RI function does) Unless it is explicitly set or overriden, such target unit would have to be that of the first quantity to preserve backward compatibility with JSR 363, we don't have to care about 275 but we must not break everything 363 did, that would cause the EC to veto the new version ;-)


Why do you say the result would be` 276.15°C `, shouldn't it be `K`?

**Comment by andi-huber** [comment]
Do we have a test-case or example code, that demonstrates the issue of **Current implementations do not obey to arithmetic laws**? I'm curious, because I don't think that this is true, once you have done all the unit conversions correctly. Or maybe I do miss the point.

**Comment by keilw** [comment]
Yes, there are plenty of test cases in the RI: https://github.com/unitsofmeasurement/indriya/blob/master/src/test/java/tech/units/indriya/function/QuantityFunctionsTemperatureTest.java. And the RI (Indriya) or Java SE 8 port (uom-se) both do this correctly in the Lambda version. Because a "target unit" like `sum(Unit)` offers is not present in the current version of `AbstractQuantity`, e.g. offered as an extra argument to `add(Quantity)` the result of `Quantity.add()` may not be controlled the same way as the Lambda function. So there is a discrepancy and mismatch in this case, but using the right operation for such need everything is there since the first Java SE 8 compliant implementation.

**Comment by desruisseaux** [comment]
@keilw: there is no problem in specifying that the target unit should be the one of the first operand. This is orthogonal with the requirement to perform calculation in system units.

> Why do you say the result would be 276.15°C, shouldn't it be K?

Computing 1°C + 2°C step by step:

* Applying the _"All operations on quantities are performed on their values stated in their system unit (SI)"_ rule, we have 1°C + 2°C = 274.15 K + 275.15 K = 549.3 K.
* Applying the _"Target units should be the unit of the first operand rule"_, we convert K to °C: 549.3 K = 276.15°C.

Note: in thermodynamic, adding temperatures in K is directly proportional to adding kinetic energy, so it has some meaning. But let keep physical meaning aside for now and stick to math only.

**Comment by desruisseaux** [comment]
@andi-huber could you elaborate on _"once you have done all the unit conversions correctly"_? Do you mean that in an expression like 1°C + 1K, it is user's responsibility to convert both operands to the same unit before to invoke `Quantity.add(Quantity)`?

**Comment by andi-huber** [comment]
@desruisseaux hmm, I guess you just answered my question.

And yes, if it would be the user's responsibility, arithmetic laws are all well fulfilled, right?

**Comment by desruisseaux** [comment]
Yes, if it is user responsibility to convert himself the quantities in the same units, then arithmetic laws are trivially fulfilled as described in the scenario 1 that you provided in [previous post](https://github.com/unitsofmeasurement/unit-api/issues/95#issuecomment-398012488). But then, what would be the value of `Quantity.add(Quantity)` compared to `double + double`? And should we throw an exception if the units are not the same?

**Comment by andi-huber** [comment]
We could consider, to leave conversion to the users responsibility only in such cases, where shifted units are involved.

**Comment by keilw** [comment]
That makes sense, with "shifted units" you mean the particular case of `Temperature`? Which brings us back to somehow recognizing which they are. Only the `Quantity` type class seems insufficient, as they can both be `Ratio` or `Interval`.

**Comment by andi-huber** [comment]
Thanks @keilw and @desruisseaux !

To elaborate more on this idea, I'm trying to picture how the use-case of adding 2 temperature quantities 1°C + 1K  could look like.

```
Quantity<Temperature> t1 = Quantities.getQuantity(1d, CELSIUS); 
Quantity<Temperature> t2 = Quantities.getQuantity(1d, KELVIN);

t1.add(t2); // produces WARNING: 'TransformedUnit addition not well defined', while acting backwards compatible
```

To resolve this WARNING, give users 4 options by having them explicitly pick a conversion (This is not a concrete naming proposal for the API, but to demonstrate the point):
```
t1.add(convertToCelsius(t2)); // 1°C + ΔT(1K) = -271.15°C
t2.add(convertToKelvin(t1)); // 1°K + ΔT(1°C) = 275.15K
t1.add(convertToCelsiusDelta(t2)); // 1°C + ΔT(1K) = 2°C
t2.add(convertToKelvinDelta(t1)); // 1°K + ΔT(1°C) = 2°K
```

That's all we would need to resolve arithmetic inconsistencies. No additional complexity required.

But this of course - which might be considered as a downside - involves:
`convertToKelvin[20°C] + convertToKelvin[5°C] ≠ convertToKelvin[25°C]`

**Comment by keilw** [comment]
Thanks @andi-huber for the examples. A warning like that could only be given through java.util.logging inside the RI, somewhat similar to one issued when running all Indriya unit tests:
```
Jun 18, 2018 5:53:02 PM tech.units.indriya.format.EBNFFormatTest testParseInverses
WARNING:  javax.measure.format.MeasurementParseException: tech.units.indriya.internal.format.TokenException:  parsing one
```
There is no way to throw a "Warning" Exception. It is something I'd be quite uncomfortable throwing a RuntimeException here, while a warning in the console (or different log setup) sounds perfectly fine.

If we found methods like `convertToCelsius()` etc. were the best approach, this must be in the RI only. For JSR 310 we already got a similar package `tech.units.indriya.quantity.time`.
Doing something similar e.g. `tech.units.indriya.quantity.temperature`, sounds fair, otherwise it must be much more generic than `toCelsius` or toKelvin, given e.g. `Acidity` also has a similar problem, but if we do this in a unit-specific way, then it is equally possible to define `toPH` or whatever in a custom extension module like uom-systems.

**Comment by andi-huber** [comment]
@keilw thanks! As I tried to make clear, this is not meant to be a concrete proposal, just a hint in one possible direction. (I agree, we would need some abstraction. And I also did not mention multiplication and how to implement a sum function, that takes multiple quantities as an input.)

Other solutions as proposed in this thread do suggest an adoption of the arithmetic and/or do involve additional state regarding quantity instances. If we can avoid this, maybe we should explore this option. If we can't, well then we can't. But I agree with 'arithmetic consistency' being a requirement.

**Comment by keilw** [comment]
As long as we can avoid a "proxy" for `Quantity` for special cases, which would cause extreme pain in many places like formatting, parsing, etc. I am open to suggestions and maybe PRs when some of you feel ready to propose them.

**Comment by desruisseaux** [comment]
> with "shifted units" you mean the particular case of Temperature? Which brings us back to somehow recognizing which they are.

Actually this is not only shifted units. This apply to all units defined by a `UnitConverter` in which `isLinear()` return `false` (with `isLinear()` defined as in JSR-363, i.e. actually an `isScale()` or `isRatio()` semantic). Temperature is only the most emblematic case.

Method like `convertToCelsius` and `convertToCelsiusDelta` would have to be generalized to every above-cited units. This bring us back to the `UnitConverter.convertDelta(double)` method proposed earlier in this thread. But let keep that aside for now - first I would like to bring more arithmetic use cases:

### Computing average temperature
This is one case where adding temperatures is a valid operation: (20°C + 30°C) / 2 = 25°C. User should be able to perform this operation without warnings, since it is perfectly valid. With Jean-Marie proposal (all units converted to system units before calculation), it still work:

`(20°C + 30°C) / 2` =
`(293.15 K + 303.15 K) / 2` =
`593.3 K / 2` = `298.15 K` = `25°C`

The key point is that the `/ 2` operation must be done in Kelvin units too, in agreement with Jean-Marie rule. The conversion to Celsius must be only the very last step. That way, the math are consistent.

### Rearranging equations
According thermodynamic, average kinetic energy is _E = kT_ where _k_ is the Boltzman constant and _T_ is the temperature in Kelvin. Let say we want to compute the total energy of two bottles at temperature T₁ and T₂, each having _N_ molecules.

* E₁ = N\*k\*T₁
* E₂ = N\*k\*T₂
* E₁ + E₂ = (N\*k\*T₁) + (N\*k\*T₂)

For performance reasons, the user may want to rearrange the equation as below. This rearrangement reduces the number of multiplications (the user can do that only if he can rely on our `Quantity` operations to be compliant with arithmetic laws):

* E₁ + E₂ = (N\*k) * (T₁ + T₂)

So this is another use case where adding temperature is a valid operation. Note also that the two bottles could be given with temperatures in °C and converting those temperature in K is the right thing to do before to apply those multiplications.

**Comment by keilw** [comment]
Remember, you also proposed https://github.com/unitsofmeasurement/unit-api/issues/39 about isLinear().

**Comment by dautelle** [comment]
> The conversion to Celsius must be only the very last step.

Conversions can be performed "on the fly" when the value is accessed. 
Since quantities operations don't need such conversion (except for output), you save a lot of time by not doing any conversion.

**Comment by desruisseaux** [comment]
@dautelle: yes, as said before the computation shall be done **_as if_** all quantities where converted to system units for calculation purpose, then the final result converted to target units in the very last step. Implementations are free to avoid conversions, provided that the final result is the same (ignoring rounding errors).

We are trying to describe the API specification. **Conceptually** we perform all those conversions. **In reality** implementations can choose whatever strategy that produce the same numerical result.

**Comment by desruisseaux** [comment]
keilw: yes I remember #39. This is why I insisted in my comment about "`isLinear()` **as currently defined**". Renaming this function is a separated issue.

**Comment by keilw** [comment]
We would have to deprecate one to rename it, that's the only way to evolve a standard.

**Comment by keilw** [comment]
Thanks for everyone who used the Doodle so far. @andi-huber, @otaviojava would you also be available to join in the evening this week: https://doodle.com/poll/gvnu2sh7wzk4yp8t ?
Right now, Thu, June 21 has the only match of all 4 who voted for a time.

**Comment by desruisseaux** [comment]
Added a wiki page with step-by-step discussion points for verifying that we have an agreement on the mathematical part (API change, if desired, would be a separated discussion): https://github.com/unitsofmeasurement/indriya/wiki/Arithmetic-operations-on-Quantity

**Comment by andi-huber** [comment]
@desruisseaux excellent write up! I much prefer to discuss one piece of concise wiki page over contributing to a comment ping-pong.

I was wondering, do you have a strong argument (Chapter 4) to emphasize, as to why commutativity of addition as in `Quantity.add(Quantity)` should be a requirement? 

Then, if we want to guarantee commutativity of addition for any arbitrary quantities, we need to consider the following ...

We do have 4 basic building blocks to make an Unit instance from (either by direct adoption or by arbitrary composition), these are:

> MULTIPLY ... MultiplyConverter x -> x * factor
> ADD ... AddConverter x -> x + offset
> LOG ... LogConverter x -> log(base, x)
> EXP ... ExpConverter x -> base^x

What ever arithmetic we design behind the scenes, we need to make sure, that we also know how to do addition and multiplication with quantities of such (possibly composed) Units.

**Comment by andi-huber** [comment]
> About the 4 basic building blocks to make Unit instances: the proposed rule ("performs calculation as if all quantities were converted to system unit") is equivalent to erasing the problem of different units of measurement and performing all arithmetic with plain double values, which is why I'm confident that it is arithmetically safe.

I agree: Given a non-system unit that is derived by a transformation f from a system unit via f: x -> f(x), then as long as we can invert this transformation, we are safe. Inversions for any of our 4 building-blocks do exist as well as for any arbitrary composition of these building blocks. (At least thats what I think. )

**Comment by andi-huber** [comment]
@desruisseaux would you mind mentioning the excellent thermodynamic example in the wiki page (chapter 4)

> E₁ + E₂ = (N*k*T₁) + (N*k*T₂)

For this formula to hold, one would need to convert T₁ and T₂ to Kelvin. Seems intuitive, no pitfalls.
> E₁ + E₂ = (N*k) * (T₁ + T₂)

Current state of API: For this formula to hold, one would need to convert T₁ and T₂ to Kelvin before the addition, not after. 

Proposed API: For this formula to hold, one is free to choose whether to convert T₁ and T₂ to Kelvin before the addition or after.

---

And adding to this, I believe there is an implicit arithmetic rule we should not break as well:

Let p an q be Quantities 'of same type' (eg. celsius, kelvin or mixed), then this rule should hold:

`p + (-1 * q) = p - q` 

also having the consequence, that 2°C - 1°C = -272.15°C (=1K)

**Comment by desruisseaux** [comment]
Thanks Andi. I added the thermodynamic example is section 4, plus a note on section 7. Also added the `p + (-1 * q) = p - q` requirement to #99 (actually I also though about that requirements).

> For this formula to hold, one would need to convert T₁ and T₂ to Kelvin. Seems intuitive, no pitfalls.

Yes, but with current specification it is unclear what an implementation would do. With current state of the API, an implementation could multiply 20°C as-is, without converting to 293.15 K. This is why the proposed requirement to (conceptually) convert everything to system units is not only arithmetically consistent, but also physically consistent regarding thermodynamics laws. But we can leave thermodynamics aside since I suspect it would create another debate; the arithmetic aspects alone should be sufficient for getting there.

**Comment by filipvanlaenen** [comment]
It's good to try and discuss this point by point, otherwise we'll never come to a conclusion.

However, I think the crux of the problem is in section 1. What we've found out is that 1 °C ≠ 1 °C, and this should be handled. I think that quantity equality should be defined in terms of value equality, unit equality, and context equality (absolute vs. increment vs. unspecified).

**Comment by andi-huber** [comment]
Thanks @desruisseaux !

> with current specification it is unclear what an implementation would do. With current state of the API, an implementation could multiply 20°C as-is, without converting to 293.15 K.

Ok, I didn't know that.

**Comment by andi-huber** [comment]
> I think that quantity equality should be defined in terms of value equality, unit equality, and context equality (absolute vs. increment vs. unspecified).

@filipvanlaenen can we be sure that a quantity's 'context' is immutable, meaning that an instance of a quantity will never ever be used in 2 conflicting contexts?

**Comment by keilw** [comment]
Of course with Multiplication there are also `multiply(double)`, how would the Quantity know another unit than the one it was created with here? In case of `20°C * 2 K` two different units are available, currently the first one wins.

**Comment by desruisseaux** [comment]
Hello @filipvanlaenen. Yes you make a good point, and I also though to that :-) - more on it below. Actually point 1 is only an attempt to clarify what we mean with "equivalence" defined in point 2. All the remaining discussion on that page use "equivalence" instead of "equality", but I agree that it may be confusing and I have no problem in rewording those definitions.

About 1 °C ≠ 1 °C, the key is in the middle of the following statement: "Two quantities A and B are said equivalent if, after conversion to the same unit of measurement by a call to `Quantity.to(Unit)`, their numerical values are the same". The key is in reference to `Quantity.to(Unit)` in that definition, and how that method performs the conversion. If one `Quantity.to(Units.KELVIN)` performs the conversion in such a way that 1°C = -272.15 K and another `Quantity.to(Units.KELVIN)` performs the conversion in such a way that 1°C = 1 K, then those quantities are not equivalent.

If a `Quantity` is 1°C, then current API allows only one behavior of `Quantity.to(Unit.KELVIN)` (unless we define a new unit of measurement named `DELTA_CELSIUS` in addition of `CELSIUS`). Which is why the 1 °C ≠ 1 °C situation is not allowed with the proposed definition of quantity equivalence. However we **could** allow `Quantity.to(Units.KELVIN)` to produce different results if the `Quantity` instances know whether their 1°C value is a measurement of an increment, thus allowing the 1 °C ≠ 1 °C situation. However this would require an API change, which I try to keep as a separated discussion (for now). My steps are:

1. First, let realize what are the consequences of arithmetic consistency without API change.
2. Next, discuss if the group accepts API changes for addressing those consequences.
3. Finally, discuss what could be those API changes (there is various possibilities: subtype, enumeration…).

**Comment by desruisseaux** [comment]
> can we be sure that a quantity's 'context' is immutable, meaning that an instance of a quantity will never ever be used in 2 conflicting contexts?

The context equality (absolute vs. increment vs. unspecified) would be - if I understood correctly @filipvanlaenen intent - a new property. This bring us to API change again. Whether it should be a `Unit` property or a `Quantity` property is another discussion - the crux is that an information is needed which is not provided by current API.

**Comment by andi-huber** [comment]
>  ... This bring us to API change again. ...

@desruisseaux I agree.

**Comment by desruisseaux** [comment]
> In case of 20°C * 2 K two different units are available, currently the first one wins.

With the proposed rules, 20°C would have to be converted to 293.15 K before the multiplication. While I have no example in mind of thermodynamic formulas involving a multiplication of 2 temperatures, I suspect that conversion to Kelvin would be mandatory anyway for being physically valid.

**Comment by keilw** [comment]
It'll need some information to enforce such rule, otherwise it could only be picked by a certain quantity type (keep in mind, Java currently has no knowledge of the Temperature at runtime because of type erasure!) while not converting it for others. It would be ludicrous to force a calculation of `1 foot` and `6 inch` to `metre` and lose both original non-SI units.

**Comment by desruisseaux** [comment]
@keilw: there is no need to know the `Quantity` type. The `Unit` instance provides all needed information.

**Comment by keilw** [comment]
@desruisseaux Assuming no API change, what would you use in `Unit`? 
`Dimension` does not really have a name or symbol, but there is a distinct instance for `TEMPERATURE`.
Or `getSystemUnit()` which in those cases normally would be expected to be `KELVIN`.

**Comment by andi-huber** [comment]
I also think, there's no API modification required ...

To express in pseudo-code (overly simplified) addition arithmetic:
```

Quantity a ...;
Quantity b ...;

// calculate result = a + b 
Quantity resultInBaseUnit = convertToBaseUnit(a).add( convertToBaseUnit(b) ) ;

// let unit of a 'win', that's gonna be the unit of the result
Quantity result = resultInBaseUnit.to( a.getUnit() );

// some magic function
Quantity convertToBaseUnit(Quantity x) {
	Map baseUnitOfX = x.getUnit().getBaseUnit();
	...
	return convertedQuantity;
}
```

**Comment by filipvanlaenen** [comment]
I think we need a change to the API, because the current API doesn't fully capture all the information needed to perform temperature calculations.

> With the proposed rules, 20°C would have to be converted to 293.15 K before the multiplication.

Or 20 K, because it can depend.

> While I have no example in mind of thermodynamic formulas involving a multiplication of 2 temperatures,

Which could be a multiplication of two absolute temperatures, or a multiplication of an absolute temperature with a temperature increment, or a multiplication of two temperature increments.

To give an example of the first and the last one: a geometric average would require a multiplication.

> I suspect that conversion to Kelvin would be mandatory anyway for being physically valid.

Unless we're dealing with temperature increments, in which case it doesn't matter.

**Comment by desruisseaux** [comment]
The proposed API changes are for determining if 1°C is a measurement (274.15 K) or an increment (1 K). Let keep that discussion aside for now and look only at the proposal for current API, which is:

> Calculation shall be done as if all quantities were converted to system units before calculation.

An implementation could apply this rule in a straightforward way (omitting parameterized types since they are not needed for this discussion). Following implementation example is inefficient, but this is not the point of this discussion.

```java
Quantity add(Quantity that) {
    Unit u1 = this.getUnit();
    Unit u2 = that.getUnit();
    Unit system = u1.getSystemUnit();
    if (!system.equals(u2.getSystemUnit())) {
        // Should never happen if parameterized types were enforced.
        throw new MeasurementException("Incompatible units");
    }
    double val1 = this.to(system).getValue();
    double val2 = that.to(system).getValue();
    Quantity result = new QuantityOfTheSameType(val1 + val2, system);
    result = result.to(u1);    // This conversion is optional; any equivalent unit is okay.
    return result;
}
```

For performance reason, an implementation may want to optimize the process and avoid the conversion to `system` unit when possible. Such optimization is arithmetically equivalent to above code if both `u1 `and `u2` are convertible to system unit with a converter where `UnitConverter.isLinear()` returns `true`. In all other cases, conversions to system units is mandatory. But this is only an optimization; the key point is that **conceptually** the result shall be as if above code was executed.

The tendency to put `Quantity` type in the discussion may be based on the belief that this problem is specific to temperature. It is not - this problem is independent of the quantity type. This problem is entirely related to the nature of the mathematical formula relating a unit to its system unit, no matter what is the quantity.

**Comment by desruisseaux** [comment]
@filipvanlaenen yes I also think that an API change is desirable and I agree with all the points you raised. I defer the discussion on API change for now just for making sure that the current situation is well understood before to move to that next step.

**Comment by keilw** [comment]
@desruisseaux Anybody could raise a PR like the above almost instantly without API change.

I find it OK with one exception:
```
   ...
    result = result.to(u1);    // This conversion is **MANDATORY**; Otherwise we break backward compatibility with JSR 363 in 99% of all cases or more, even 1 cm+1 mm would be mangled into m return result;
}
```

**Comment by andi-huber** [comment]
For me, with 2 days into this discussion, it seems clear now, that 
> Calculation shall be done as if all quantities were converted to system units before calculation.

a) do-able without API change and b) consistent regarding arithmetic.

Meaning I'm ready for the next step. (... the consequences)

**Comment by desruisseaux** [comment]
> Anybody could raise a PR like the above almost instantly without API change.

* With the code example that I posted, 1°C + 2°C = 276.15°C.
* This code does not handle the cases raised by @filipvanlaenen, where 2°C would be an increment for example (so that 1°C + 2°C = 3°C). Supporting those cases require an API change.
* No matter what we do, the Javadoc needs to be more explicit about how arithmetic operations are performed if we want equivalent result across different implementations.

> `result = result.to(u1);`    // This conversion is MANDATORY (…snip…)

It is optional on an arithmetic point of view. It may be mandatory on a backward compatibility point of view, but this is an orthogonal issue. This discussion tries to keep mathematic and existing implementations as two separated issues (as long as they are not conflicting).

**Comment by keilw** [comment]
IC, especially the Spec Leads among us have to ensure, the compatibility is not broken unless unavoidable for a very special reason (e.g. °C/K) but otherwise there should be no massive code changes if someone updates from 1.0 to 2.0.

**Comment by filipvanlaenen** [comment]
> Calculation shall be done as if all quantities were converted to system units before calculation.

I agree, but this statement doesn't imply that 1 °C + 2 °C = 276.15 °C, because even 1 °C + 2 °C = 3 K isn't inconsistent with it.

Therefore

> is a) do-able without API change 

isn't true. Either the API will provide a result that in the majority of the use cases (2 out of 3) is incorrect, or it would require the user to do the relevant conversion outside the mathematical operation.

Note that changing the behaviour of a method is an API change too, even if you don't change the method signature, and the changes proposed higher up can hardly be called bug fixes.

**Comment by desruisseaux** [comment]
> I agree, but this statement doesn't imply that 1 °C + 2 °C = 276.15 °C, because even 1 °C + 2 °C = 3 K isn't inconsistent with it.

The statement is incomplete. Let me complete it:

> Calculation shall be done as if all quantities were converted to system units by a call to `Quantity.to(Unit)` before calculation.

The key is "by a call to `Quantity.to(Unit)`". With current API, `Quantity.to(Unit)` has no way to know if it should convert 2°C into 2 K or 275.15 K, unless we invent a new `CELSIUS_DELTA` unit. Allowing `Quantity.to(Unit)` to know if it is converting a _measurement_ or a _difference_ is the purpose of the API change.

**Comment by filipvanlaenen** [comment]
OK. I don't know where the following statement was made originally,

> Calculation shall be done as if all quantities were converted to system units by a call to Quantity.to(Unit) before calculation.

but it certainly contains an API change since it would change the meaning of some of the methods in the current API.

**Comment by desruisseaux** [comment]
Yes, so we could said that an API change is required anyway, and the question is whether the change include addition of new properties or methods (I would said it should).

**Comment by keilw** [comment]
I think so, too, it would be more transparent. Adding the unit to convert to and store as could be one option (like those Lambda/library functions also do) or a concept of "measurement level" which Wikipedia also mentions and while contested by some experts the 4 main levels look mature and stable since the end of WW2.

**Comment by andi-huber** [comment]
When stating 'is do-able without API change' yesterday, what I meant was, that from the code perspective, no new interfaces or methods would need to be introduced. But I agree, that the behavior and hence associated Java-doc would have to change, which strictly speaking one could also see as a change in API, yes.

**Comment by desruisseaux** [comment]
> Adding the unit to convert to and store as could be one option

No, the problem is not that we need an extra option for the unit to convert. The target unit (not to be confused with the computational unit) does not matter, as long as the quantities are _equivalent_ as defined in the wiki page.

The problem is to know if a quantity is a _measurement_ of a _difference_ (or increment). This is orthogonal with unit. It could be a sub-type of `Quantity`, or a property in `Quantity` **in complement** with the `Unit`. A property directly in `Unit` has also been suggested.

**Comment by keilw** [comment]
That with e.g. some `enum` would be the least invasive and most definition compliant (i.E. https://en.wikipedia.org/wiki/Level_of_measurement, we quote Wikipedia in other places, too) addition to the API.

**Comment by desruisseaux** [comment]
To summarize where we are now: this issue was about clarifying arithmetic operations on `Quantity`. I think we have reached an agreement that the only arithmetically consistent way before introducing new types/properties/methods is to require the [rule stated above](https://github.com/unitsofmeasurement/unit-api/issues/95#issuecomment-398523634), despite its counter-intuitive consequences. Then I propose the following actions:

1. Update `Quantity` Javadoc with:
   * definition of quantity equivalence
   * above rule for arithmetic operations
2. Update implementations accordingly.
3. Close this issue if peoples agree with the javadoc.
4. Create a new issue for new types, properties or methods to add to the API.

**Comment by filipvanlaenen** [comment]
I don't think this is such a good idea. I would leave things like they are for now, and make the change in one API change instead of two. Also, I'm not so fond of API changes that go completely under the radar, i.e. a change in behaviour without a warning. There may be users out there for which the current implementation works just fine, and then suddenly, they're getting results that don't make much sense, without a compilation error or a runtime exception. As a user, I wouldn't be particularly happy with that.

**Comment by desruisseaux** [comment]
In my understanding, the API change - at least for the part standardized by JSR - would have to be done all in one shot, since JSR-385 is about "Unit API 2.0" and must be voted by JCP (Oracle and other members) before to be released.

However implementations may have as many releases as they want. So it is possible for Indriya to take a progressive approach if they want, but all Indriya official releases have to be on JSR-363 API as long as JSR-385 has not been officially approved (but Indriya may have development releases on JSR-385 API for testing purpose).

JSR-363 API said nothing about the behavior of `Quantity` arithmetic operations. So a change of behavior of Indriya in this aspect would not be a compatibility break with JSR-363. It may be a compatibility break with previous Indriya versions, but not with the specification. So I think it is up to Indriya project to manage this migration as they wish; I don't think it would interfere with JSR-385.

**Comment by keilw** [comment]
I think it's a good idea to create detail tasks where agreed on. Indriya did not exist until the new JSR was started. And Indriya 1.0 is a fully compatible variation of the Java 8 implementation that was not the RI earlier. It is a bridge to the new RI, actually 1.0 is still using the old "tec.*" namespace, so after Indriya 1.2 there are package changes, but the general structure is compatible. The only things that are going to change are e.g. `MetricPrefix` moving to the API and a few concrete classes like the Pi*Converter ones. Although the TCK and JCP EC likely won't care about such details, users and projects or commercial products (OpenHAB/SmartHome is the foundation for commercial SmartHome products like "Magenta SmartHome" by Deutsche Telekom) do care about some predictability, so suddenly returning only SI units where a different unit passed as arguments to an operation was returned before is an absolute no-go. If conversion happens back and forth under the hood, that's a different story, but the overall behavior should not be broken across the RI.

**Comment by keilw** [comment]
Closing this after the creation of #130 and https://github.com/unitsofmeasurement/indriya/issues/128

**Comment by keilw** [comment]
@kaikreuzer @htreu it took a while to apply that, but we just finished #130 and thanks to #131 there is a way to calculate e.g. temperatures (with an `INTERVAL` level of measurement) differently than other units.
